daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/unparsers/UState.scala

```diff
@@ -344,7 +342,21 @@ abstract class UState(
       // Finished means you won't add data to the end of it any more.
       // It does NOT prevent information like the absoluteBitPos to
       // propagate.
-      dos.setFinished(this)
+      //
+      // Need this clone or we could possibly end up in an infinite loop inside of
+      // DirectOrBuffereDataOutputStream.cleanUp()
```

Comment:
I think there's a better explanation for why we need a clone of the UState, maybe something like this:

> When setFinished is called, the DOS is going to store the state that we pass into it in finishedFormatInfo. At some point in the future, this DOS will become a direct DOS that may be delivered to a following buffered DOS. When that happens, this saved finishedFormatInfo will be used. However, this all requires that the UState does not change while we are waiting for this DOS to become direct. If the state does change, it will be incorrect and can lead to undefined behavior. To prevent this, we must clone the UState so it can no longer change, and pass that clone into setFinished.